src/Provider/AbstractProvider.php

```diff
@@ -370,20 +360,60 @@ protected function getAccessTokenQuery(array $params)
     }
 
     /**
-     * Checks that a provided grant is valid, or attempts to produce one if the
-     * provided grant is a string.
+     * Returns the full URL to use when requesting an access token.
      *
-     * @param mixed $grant
-     * @return AbstractGrant
+     * @param array $params Query parameters
      */
-    protected function verifyGrant($grant)
+    protected function getAccessTokenUrl(array $params)
     {
-        if (is_string($grant)) {
-            return $this->grantFactory->getGrant($grant);
+        $url = $this->getBaseAccessTokenUrl();
+
+        if ($this->getAccessTokenMethod() === 'GET') {
+            $query = $this->getAccessTokenQuery($params);
+            return $this->appendQuery($url, $query);
         }
 
-        $this->grantFactory->checkGrant($grant);
-        return $grant;
+        return $url;
+    }
+
+    /**
+     * Returns headers that are only used when requesting an access token.
+     *
+     * @return array
+     */
+    protected function getAccessTokenHeaders()
+    {
+        return $this->getDefaultHeaders();
+    }
+
+    /**
+     * Returns the request body for requesting an access token.
+     *
+     * @param  array $params
+     * @return string
+     */
+    protected function getAccessTokenBody(array $params)
+    {
+        return http_build_query($params);
```

Comment:
This is a distinction from `getAccessTokenQuery` because the URL and POST body may not always have the same parameters and encoding.